ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +64,221 @@ data NtpStatus =
-mkSockets tracer settings =
-    doMkSockets `catch` handleIOException >>= \case
-        Just socks -> pure socks
-        Nothing     -> do
-            traceWith tracer NtpTraceMkSocketsNoSockets
---            logWarning "Couldn't create either IPv4 or IPv6 socket, retrying in 5 sec..."
-            threadDelay 5000000
-            mkSockets tracer settings
+        runProtocol _version _localAddr [] = return []
+        runProtocol _version Nothing    _  = return []
+        runProtocol version (Just addr) servers = do
+            socketAction tracer ntpSettings addr servers >>= \case
+                Left err -> do
+                    traceWith tracer $ NtpTraceRunProtocolError version err
+                    return []
+                Right [] -> do
+                    traceWith tracer $ NtpTraceRunProtocolNoResult version
+                    return []
+                Right r@(_:_) -> do
+                    traceWith tracer $ NtpTraceRunProtocolSuccess version
+                    return r
+
+socketAction ::
+       Tracer IO NtpTrace
+    -> NtpClientSettings
+    -> AddrInfo
+    -> [AddrInfo]
+    -> IO (Either IOError [ReceivedPacket])
+socketAction tracer netSettings localAddr destAddrs
+    = tryIOError $ bracket acquire release action
   where
-    doMkSockets :: IO (Maybe Sockets)
-    doMkSockets = do
-        addrs <- udpLocalAddresses
-        (<>) <$> (createAndBindSock tracer IPv4 addrs)
-             <*> (createAndBindSock tracer IPv6 addrs)
-
-    handleIOException :: IOException -> IO (Maybe Sockets)
-    handleIOException e = do
-        traceWith tracer $ NtpTraceMkSocketsIOExecption e
-        threadDelay 5000000
-        doMkSockets
+    acquire :: IO Socket
+    acquire = do
+        s <- Socket.socket (addrFamily localAddr) Datagram Socket.defaultProtocol
+        traceWith tracer NtpTraceSocketOpen
+        return s
+
+    release :: Socket -> IO ()
+    release s = do
+        Socket.close s
+        traceWith tracer NtpTraceSocketClosed
+
+    action :: Socket -> IO [ReceivedPacket]
+    action socket = do
+        Socket.setSocketOption socket ReuseAddr 1
+        Socket.bind socket (addrAddress localAddr)
+        inQueue <- atomically $ newTVar []
+        _err <- withAsync (send socket  >> loopForever)      $ \sender ->
+               withAsync timeout    $ \delay ->
+               withAsync (reader socket inQueue ) $ \revc ->
+                    waitAnyCancel [sender, delay, revc]        
+        atomically $ readTVar inQueue
+
+    send :: Socket -> IO ()
+    send sock = forM_ destAddrs $ \addr -> do
+        p <- mkNtpPacket
+        err <- tryIOError $ Socket.ByteString.sendManyTo sock
+                          (LBS.toChunks $ encode p) (setNtpPort $ Socket.addrAddress addr)
+        case err of
+            Right _ -> traceWith tracer NtpTracePacketSent
+            Left e  -> do
+                traceWith tracer $ NtpTracePacketSentError e
+                ioError e
```

Comment:
I would use a `onException` (or `catch` if you only want to act on `IOException`s`).  This will make transition to `MonadCatch` easier (it does not have the low level `System.IO` exception control flow functions).

Reply:
lets leave that for a new PR